File: ObserverV2/src/ClockTimer.cpp
#include "../src/Notifier.cpp"

class ClockTimer: public IClockTimer
{
    public:
        ClockTimer(ISubscriber* subscriber):refCount(0),notifier(new Notifier(subscriber)) {
            acquire();
            m_hours = 0;
            m_minutes = 0;
            m_seconds = 0;
            std::cout << "ClockTimer::ClockTimer" << std::endl;
            this->subscriber = subscriber;
        }

        virtual ~ClockTimer(){
            std::cout << "ClockTimer::~ClockTimer" << std::endl;
            notifier->release();
        }

        int getHour( void ) {
            return this->m_hours;
        }

        int getMinute( void ) {
            return this->m_minutes;
        }

        int getSecond( void ) {
            return this->m_seconds;
        }

        ISubscriber* getSubscriber() {
            return this->subscriber;
        }

        void tick () {
            m_seconds++;
            if(m_seconds == 60) {
                m_seconds = 0;
                m_minutes++;
                if(m_minutes == 60) {
                    m_seconds = 0;
                    m_minutes = 0;
                    m_hours++;
                    if(m_hours == 24) {
                        m_seconds = 0;
                        m_minutes = 0;
                        m_hours = 0;
                    }
                }
            }
            notifier->notify();
        }

        //IMemoryManagement interface implementation

        void acquire(){
            refCount++;
        }

        void release(){
            if(--refCount == 0)
                delete this;
        }

    private:
        int refCount;
        int m_hours;
        int m_minutes;
        int m_seconds;
        ISubscriber* subscriber;
        INotifier* notifier;
        
};

File: ObserverV2Java/ObserverJava/src/src/ClockTimer.java
package src;

import include.IClockTimer;
import include.INotifier;
import include.ISubscriber;

public class ClockTimer implements IClockTimer {

    public ClockTimer(ISubscriber subscriber) {
        notifier = new Notifier(subscriber);
        hours = 0;
        minutes = 0;
        seconds = 0;
        this.subscriber = subscriber;
    }

    public int getHour() {
        return this.hours;
    }

    public int getMinute() {
        return this.minutes;
    }

    public int getSecond() {
        return this.seconds;
    }

    public ISubscriber getSubscriber() {
        return this.subscriber;
    }

    public void tick() {
        seconds++;
        if (seconds == 60) {
            seconds = 0;
            minutes++;
            if (minutes == 60) {
                seconds = 0;
                minutes = 0;
                hours++;
                if (hours == 24) {
                    seconds = 0;
                    minutes = 0;
                    hours = 0;
                }
            }
        }
        notifier._notify();
    }


    private INotifier notifier;
    private int hours;
    private int minutes;
    private int seconds;
    private ISubscriber subscriber;
}
